### PR TITLE
auth: geoip - Fix static lookup when using weighted records on multiple record types

### LINKED ISSUE #7051: GeoIP sometimes does not answer on when using weighted records and different records types
- Program: Authoritative
 - Issue type: Bug report

### Short description

Hello, when using the GeoIP backend with weighted records and different types (A and AAAA) the server respond nothing half of the time (the sum of the weight of the A records is the same as the AAAA records)

### Environment

 - Operating system: CenOS 7.4.1708
 - Software version: 4.1.4
 - Software source: PowerDNS repository

### Steps to reproduce

Configuration:

```
setgid=pdns
setuid=pdns
launch=geoip,bind

bind-config=/etc/pdns/bind.conf
do-ipv6-additional-processing=no
edns-subnet-processing=yes
geoip-database-cache=memory
geoip-database-files=/usr/share/GeoIP/GeoIPCity.dat
geoip-zones-file=/etc/pdns/geoip.yaml
local-port=53
```

geoip.yaml

```---
domains:
- domain: example.com
  ttl: 300
  records:
    example.com:
    - a:
        content: 192.168.0.1
        weight: 50
    - a:
        content: 192.168.0.2
        weight: 50
    - aaaa:
        content: 2042::1
        weight: 50
    - aaaa:
        content: 2042::2
        weight: 50
```

```
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
2042::2
# dig AAAA example.com @localhost  +short
2042::2
# dig AAAA example.com @localhost  +short
2042::2
# dig AAAA example.com @localhost  +short
2042::2
# dig AAAA example.com @localhost  +short
2042::1
# dig AAAA example.com @localhost  +short
2042::2
# dig AAAA example.com @localhost  +short
2042::1
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
2042::1
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
# dig AAAA example.com @localhost  +short
2042::2
```
### Expected behaviour

I expect to get an answer each time, 2042::1 or 2042::2.

### Other information

Reading the geoip backend here https://github.com/PowerDNS/pdns/blob/auth-4.1.4/modules/geoipbackend/geoipbackend.cc#L327, the weight is computed for each records without distinction of record types.
Perhaps it could be relevant to compute weight for each type.

### PR DESCRIPTION
**Pull request by Annih**
### Short description
Computes the record weight/probability per QType and adapt the code to avoid empty answer in a mixed-weighted-Qtype situation.

Closes #7051.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: modules/geoipbackend/geoipbackend.cc [hunk 1]
```diff
@@ -261,24 +261,30 @@ void GeoIPBackend::initialize() {
...
+        weights[rr.qtype.getCode()] += rr.weight;
         if (rr.has_weight) has_weight = true;
       }
       if (has_weight) {
         // put them back as probabilities and values..
         for(auto &rr: item.second) {
-          rr.weight=static_cast<int>((static_cast<float>(rr.weight) / weight)*1000.0);
-          sum += rr.weight;
+          uint16_t rr_type = rr.qtype.getCode();
+          rr.weight=static_cast<int>((static_cast<float>(rr.weight) / weights[rr_type])*1000.0);
```
Comment (rgacogne) [inline modules/geoipbackend/geoipbackend.cc]: This was already the case with the existing code, but it looks like we would happily divide by `0.0` if the weight of all records for a given type is set to `0`.

Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: What would be the desired behavior in that case?
Maybe the yaml parser should just ignore records with a weight of 0?


Anyway it seems out of the scope of this review, what do you think @rgacogne?

Comment (rgacogne) [reply modules/geoipbackend/geoipbackend.cc]: I think the parser should error out on a weight of `0`.

I agree it's a bit out of the scope so if you don't feel like fixing that in your PR just let me know and I'll either open a new PR or an issue so we don't forget to fix it.

Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: OK I'm doing it in another review :)

**Comment by Annih** [comment]
@rgacogne I'm new in the pdns opensource world, in order to backport this change to the *4.1.x* branch, what is the process? 

I read I need first to wait that this is one is merged, but I still have some question:
- Should I do the backport or a maintainer will do it?
- Should all the fixes be backported, or only the main fix?

**Comment by Habbie** [comment]
> * Should I do the backport or a maintainer will do it?

If you want it, it's best if you do it, instead of having to wait for us.



> * Should all the fixes be backported, or only the main fix?

We prefer minimal backports whenever possible.

**Comment by rgacogne** [comment]
Perhaps @cmouse might want to review this before we merge it?

File: modules/geoipbackend/geoipbackend.cc [hunk 2]
```diff
@@ -307,25 +313,25 @@ GeoIPBackend::~GeoIPBackend() {
...
 
   if (i != dom.records.end()) { // return static value
     for(const auto& rr : i->second) {
+      if (qtype != QType::ANY && rr.qtype != qtype) continue;
+
       if (rr.has_weight) {
         gl.netmask = (v6?128:32);
-        int comp = cumul_probability;
-        cumul_probability += rr.weight;
+        int comp = cumul_probabilities[rr.qtype.getCode()];
```
Comment (cmouse) [inline modules/geoipbackend/geoipbackend.cc]: I am bit worried that QType::ANY will no longer work here.

File: regression-tests/backends/geoip-master [hunk 3]
```diff
@@ -26,6 +26,13 @@ domains:
       - ns: ns1.example.com
       - ns: ns2.example.com
       - mx: 10 mx.example.com
+    mixed_weight.geo.example.com:
+      - a:
+          content: "%ip4"
+          weight: 10
+      - txt:
+          content: text
+          weight: 10
```
Comment (cmouse) [inline regression-tests/backends/geoip-master]: this test does not really test the functionality

Comment (Annih) [reply regression-tests/backends/geoip-master]: Could you be more specific on the "functionality" which should be tested and which is not?

Is there something unclear in the commit message where I introduce the test? If yes, how should I reword it to make it clearer?

My goal is to demonstrate we do not regress/retrigger #7051, and ensure that the ANY question answer both records, the A question return the A record and the the TXT return the TXT.
Do you think my test is covering this goal?
Do you think my test is enough?

File: modules/geoipbackend/geoipbackend.cc (hunk 2 above)
Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: Do you mean in the case where someone put a ANY record in the yaml? (I would say this is not valid)

Otherwise I would actually say that this patch fixes the behaviour with the QType::ANY. Before this patch a ANY request with mixed QType records would only return 1 record (or zero due to the bug in #7051)
i.e. as I tried to demonstrate with the new regression test, if you specify *A* and *TXT* values in a record, I would expect the ANY question to return 1 record for the *A* type and another for the *TXT* record.

File: regression-tests/backends/geoip-master (hunk 3 above)
Comment (cmouse) [reply regression-tests/backends/geoip-master]: although not sure if it's even possible to test it well.

File: modules/geoipbackend/geoipbackend.cc (hunk 2 above)
Comment (cmouse) [reply modules/geoipbackend/geoipbackend.cc]: I am wondering if this still makes sense, because now it can return "different" TXT-A pair if there are more than one specified.

Consider following example:
 - A 1.2.3.4, weight = 10
 - A 4.3.2.1, weight = 10
 - TXT "1-2-3-4", weight = 10
 - TXT "4-3-2-1", weight = 10

If I am reading this code correctly, it can end up returning four different kinds of results.

So, is the requirement to have corresponding records returned, or is it ok for them to behave independently of each other.

Seems that ANY query is handled correctly after all, I read it wrong the first round. But only question now remains is that is it OK that records are chosen randomly by type? If it's OK, then this works correctly.

Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: @Habbie @rgacogne or @cmouse who should take this decision?

Could we move forward? I'll send the backport review once this is merged.

**Comment by Annih** [comment]
@rgacogne & @cmouse any thought about this?
I know you are working hard on 4.2.0, it would be great if this change could be included in the release.

File: modules/geoipbackend/geoipbackend.cc (hunk 2 above)
Comment (cmouse) [reply modules/geoipbackend/geoipbackend.cc]: I think the result-set needs to contain one of TXT and one of A in this configuration to be correct.

Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: That's what the new implementation produces; a single random record per QTYPE.

Comment (cmouse) [reply modules/geoipbackend/geoipbackend.cc]: so can you paste what `dig any name.geo.example.com` gives?

Comment (Annih) [reply modules/geoipbackend/geoipbackend.cc]: Yes certainly I can :), this is exactly the behavior I tried to demonstrate in the added regression test with the ANY query, but due to the randomness of the result I had to simplify the data.

So with the a setup similar to what you described:
```
    name.geo.example.com:
      - a:
          content: 1.2.3.4
          weight: 10
      - a:
          content: 4.2.3.1
          weight: 10
      - txt:
          content: 1-2-3-4
          weight: 10
      - txt:
          content: 4-2-3-1
          weight: 10
```

Here are the results:
```
#1:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#2:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#3:
name.geo.example.com.   30      IN      TXT     "1-2-3-4"
name.geo.example.com.   30      IN      A       1.2.3.4
#4:
name.geo.example.com.   30      IN      TXT     "4-2-3-1"
name.geo.example.com.   30      IN      A       4.2.3.1
#5:
name.geo.example.com.   30      IN      TXT     "1-2-3-4"
name.geo.example.com.   30      IN      A       1.2.3.4
```

Interestingly the pairs are preserved, but I'm not sure this is a feature that should be advertised ^^

**Comment by aerique** [comment]
@rgacogne do you still approve and @cmouse do you approve?

**Comment by Annih** [comment]
Thanks @aerique for the reminder, it has been a while that this PR didn't move :)
@cmouse could you please tell us if you approve or not? If you don't, could you drive me on the right path to fix this patch and move forward?

**Comment by cmouse** [comment]
i already wrote in earlier code that i approve, but lets approve here too

**Comment by Annih** [comment]
Thanks @cmouse :)
So what is required to merge this, who is allowed to do it?

**Comment by aerique** [comment]
> So what is required to merge this, who is allowed to do it?

I or one of the other PowerDNS people will merge it in the coming days.
